src/boot/bootctl-uki.c

```diff
@@ -0,0 +1,111 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "bootctl.h"
+#include "bootctl-uki.h"
+#include "fd-util.h"
+#include "parse-util.h"
+#include "pe-header.h"
+
+#define MAX_SECTIONS 96
+
+static const uint8_t dos_file_magic[2] = "MZ";
+static const uint8_t pe_file_magic[4] = "PE\0\0";
+
+static const uint8_t name_osrel[8] = ".osrel";
+static const uint8_t name_linux[8] = ".linux";
+static const uint8_t name_initrd[8] = ".initrd";
+
+static int pe_sections(FILE *uki, struct PeSectionHeader **ret, uint16_t *ret_n) {
+        _cleanup_free_ struct PeSectionHeader *sections = NULL;
+        struct DosFileHeader dos;
+        struct PeHeader pe;
+        uint16_t scount;
+        uint64_t soff, items;
+        int rc;
+
+        *ret = NULL;
+        *ret_n = 0;
```

Comment:
plese don't. we have this rule in systemd that return parameters of functions are to be left untouched on failure, and are always initialized on success. But if you initialize this here, there are still plenty occasions we might fail further down so you touched the return parameters even tough we failed. Thus, please add a jump label to the bottom, maybe called "not_found" or so where you do this parameter initializaion and return 0.